R/GeospatialFunctions.R

```diff
@@ -450,28 +798,143 @@ TADA_GetATTAINS <- function(.data, return_sf = TRUE) {
           dplyr::distinct(assessmentunitidentifier, .keep_all = TRUE),
         silent = TRUE
       )
+      if(is.null(ATTAINS_polygons) || nrow(ATTAINS_polygons) == 0){
+        ATTAINS_polygons <- NULL
+      }
+      
     }))
-
+    # If there are ATTAINS catchments, fill_catchments = FALSE, and return_sf = TRUE:
     return(list(
       "TADA_with_ATTAINS" = TADA_with_ATTAINS,
       "ATTAINS_catchments" = ATTAINS_catchments,
       "ATTAINS_points" = ATTAINS_points,
       "ATTAINS_lines" = ATTAINS_lines,
       "ATTAINS_polygons" = ATTAINS_polygons
     ))
-  }
+    
+    # If there IS at least some ATTAINS data, and fill_catchments = TRUE...
+  } else if (!is.null(nearby_catchments) & fill_catchments == TRUE) {
+    # ... link WQP features to the ATTAINS catchment feature(s) they land in:
+    TADA_with_ATTAINS <- TADA_DataRetrieval_data %>%
+      sf::st_join(., nearby_catchments, left = TRUE)
+    
+    missing_attains <- dplyr::filter(TADA_with_ATTAINS, is.na(ATTAINS.submissionid))
+    
+    # Splitting up sites with and without ATTAINS, so remove those without ATTAINS:
+    TADA_with_ATTAINS <- TADA_with_ATTAINS %>%
+      dplyr::filter(!is.na(ATTAINS.submissionid))
+    
+    # If there are no WQP observations without missing ATTAINS features, return empty df for
+    # TADA_without_ATTAINS
+    if(nrow(missing_attains) == 0){
+      
+      print('All WQP features intersect an ATTAINS catchment. Returning empty dataframe for "TADA_without_ATTAINS".')
+      
+      if(resolution %in% c("Med", "med")){
+        TADA_without_ATTAINS <- tibble::tibble(NHD.comid = character(),
+                                               NHD.resolution = character(),
+                                               NHD.catchmentareasqkm = numeric())
+        nhd_catchments <- NULL
+      } else if(resolution %in% c("Hi", "hi")){
+        TADA_without_ATTAINS <- tibble::tibble(NHD.nhdplusid = character(),
+                                               NHD.resolution = character(),
+                                               NHD.catchmentareasqkm = numeric())
+        
+        nhd_catchments <- NULL
+      } else {stop('Please select between "Med" or "Hi" for your NHD resolution.')}
+    }
+    
+    # If there are some observations with no attains features, grab those sites' intersecting NHD catchments:
+    if(nrow(missing_attains) > 0){
+      
+      # Downloading NHD data to fill in missing ATTAINS features. Depending on the number of observations and
+      # their spatial extent, this can take a while.
+      nhd_catchments <- fetchNHD(.data = missing_attains,
```

Comment:
See earlier comment for `fetchNHD` - in description of the function it seems intent was to allow return of flowlines and waterbodies as well for selected resolution, but appears currently only return of catchments is built in?